src/plugins/external_tool_support/src/ExternalToolManager.cpp

```diff
@@ -193,28 +190,42 @@ QString ExternalToolManagerImpl::addTool(ExternalTool *tool) {
         foreach (const QString &dependency, toolDependencies) {
             dependencies.insertMulti(dependency, tool->getId());
         }
+    }
+}
+
+void ExternalToolManagerImpl::addToolToListsIfNecessary(const ExternalTool &tool) {
+    if (!tool.isValid()) {
+        if (tool.getDependencies().isEmpty() || dependenciesAreOk(tool.getId())) {
+            validateList << tool.getId();
+        }
+    }
+
+    if (!validateList.contains(tool.getId()) && !tool.isModule() && !tool.isValid()) {
+        searchList << tool.getId();
+    }
+}
 
-        if (dependenciesAreOk(tool->getId()) && !tool->isValid()) {
-            if (tool->isModule()) {
-                QString masterId = tool->getDependencies().first();
+QString ExternalToolManagerImpl::findToolPath(const ExternalTool &tool) {
```

Comment:
The method has very unclear contract: it returns empty value for valid tools.
This is the reason I ask many questions about if () else parts: it looks like not all variants are convered.

Reply:
Yes, it returns an empty string if tool is valid, and it should be. And all variants are covered (see empty else).